--- /usr/local/bin/rebuild_rbldns.txt	2009-03-20 23:03:57.000000000 -0700
+++ rebuild_rbldns.pl	2017-04-27 08:14:31.872915329 -0700
@@ -20,15 +20,15 @@
 #
 # Begin User Defined Section
 #----------------------------
-my $blacklist_file = "/var/lib/rbldns/dsbl/spammerlist";
-my $whitelist_file = "/var/lib/rbldns/dsbl/whitelist";
-my $rbl_domain     = "rbl.mydomain.com";
-my $mysql_user     = "dnsuser";
-my $mysql_pass     = "potato";
-my $mysql_database = "dnsrbl";
-my $mysql_host     = "localhost";
-my $datasource     = "dbi:mysql:database=$mysql_database;host=$mysql_host";
-my $temp_file      = "/var/lib/rbldns/dsbl/templist";
+my $dir            = $ENV{"HOMEDIR"} . "/" . $ENV{"CFG_NAME"};
+my $blacklist_file = "$dir/spammerlist";
+my $whitelist_file = "$dir/whitelist";
+my $rbl_domain     = $ENV{"RBL_DOMAIN"};
+my $my_cnf         = $ENV{"HOMEDIR"} . "/.my.cnf";
+my $db             = $ENV{"DB_NAME"};
+my $datasource     = "dbi:mysql:database=$db;mysql_read_default_file=$my_cnf";
+my $temp_file      = "$dir/templist";
+my $pid_file       = $ENV{"PID_FILE"};
 #----------------------------
 # End User Defined Section
 
@@ -40,6 +40,7 @@
 $pid = $$;
 &getopts("fhvV",\%Options);
 &usage if ($Options{'h'}); # then exit
+my $changed = 0;
 my $dbh;
                                                                                 
 if ($Options{"V"}) {
@@ -61,6 +62,9 @@
     print "Error: Could not connect to local MySQL database. (did password change?)\n";
     exit 1; # bad exit
 }
+if ($changed) {
+  &send_sighup($pid_file);
+}
 exit;
 
 ##########################
@@ -119,9 +123,27 @@
          }
       }
       close RBL;
-      `mv $temp_file $file`;
+      `diff -q $temp_file $file`;
+      if ($?) {
+        `mv $temp_file $file`;
+	$changed = 1;
+      }
       print "$count ips of type $type\n" if ($Options{'v'});
   } else {
       print "Failed to open $file for writing\n";
   }
 }
+
+sub send_sighup() {
+###########################################################
+# send SIGHUP signal to rbldnsd                           #
+###########################################################
+  my $pidfile = @_;
+  if (open PID, "< $pidfile") {
+    my $pid = <PID>;
+    close PID;
+
+    kill 1, $pid;
+    die "Lock held by PID $pid!\n" if $! == Errno::ESRCH;
+  }
+}
